### PR TITLE
pipeline validate, modify and exception handling

### LINKED ISSUE #252: Add utility notebook to check successful execution of Pipeline
**Is your feature request related to a problem? Please describe.**
After triggering automated workflows of the project, we need to have a way to check if the notebooks in the pipeline ran successfully. 

**Describe the solution you'd like**
We can have a utility notebook which simply checks the desired files uploaded to s3 on that date, something which can help validate that the pipeline ran and produced the desired metrics.

**Describe alternatives you've considered**
manually using the command line to check the presence of the metric files and raw data on s3

### LINKED ISSUE #254: Modify Pipeline configuration for realistic resource usage
**Describe the bug**
Kubeflow Pipeline configured to have high resource usage which is greater than the current requirement and causes cpus in cluster to be blocked.

**Expected behavior**
Pipline configured to use CPU and Memory based on usage reflected in [dashboard](https://grafana-route-opf-monitoring.apps.zero.massopen.cloud/d/2ZPZ1LrGz/tcoufal-testing-kubeflow-pipeline-info?orgId=1)

**Screenshots**
Large difference between actual usage and effective request

image

### LINKED ISSUE #255: Metric Notebooks failing to execute when data is written/read from S3
**Describe the bug**
As I am trying to run the Kubeflow [Pipeline](http://istio-ingressgateway-istio-system.apps.zero.massopen.cloud/pipeline/#/runs/details/9e2efd61-fa62-4a61-ab6e-65ab65405469) for the project, I am getting an error during the metric calculation steps
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-87f9b4cb33b3> in <module>
----> 1 failures_list = testgrid_labelwise_encoding(testgrid_data, 12)

~/ocp-ci-analysis/notebooks/data-sources/TestGrid/metrics/metric_template.ipynb in testgrid_labelwise_encoding(data, label)
    254     "    lst = []\n",
    255     "    for run_length in x:\n",
--> 256     "        extension = [run_length[\"value\"]] * run_length[\"count\"]\n",
    257     "        lst.extend(extension)\n",
    258     "    return lst\n",

IndexError: list index out of range
```
I tried to reproduce this error by running the notebooks In AUTOMATION from within JupyterHub as well and saw the same error.

When, I configured the .env to run locally (not in automation) and ran the `get_raw_data.ipynb` to download and save the data locally and run the metric notebooks on those, I no longer see the error.

**To Reproduce**
1. Clone Repo in JH
2. Set `IN_AUTOMATION=True` in `.env`
3. Run metric notebooks eg: `number_of_flakes`, `build_pass_failure`
4. See error `IndexError: list index out of range`
5. Remove `IN_AUTOMATION` flag from `.env`
6. Run `get_raw_data` notebook and download data locally
7. Run the same notebooks and they run without any error.

Instead of steps 1,2,3,4 you can also trigger the KFP pipeline to see the same error.

**Screenshots**
Screen Shot 2021-05-04 at 12 54 22 PM
Screen Shot 2021-05-04 at 12 58 30 PM

**Additional Context**
The Pipeline ran successfully a day back http://istio-ingressgateway-istio-system.apps.zero.massopen.cloud/pipeline/#/runs/details/d91463e5-d1e0-4c35-916c-62d761976645

### PR DESCRIPTION
**Pull request by oindrillac**
## Related Issues and Dependencies

In this PR,

- I am adding a validation utility notebook to check files in the S3 bucket which indicate the successful execution of a Pipeline - resolves #252 
- I am modifying the resource configurations in the pipeline as per realistic resource usage - resolves #254 
- I am adding some exception handling in `metric_template` to prevent the errors seen in #255 - resolves #255 

## This introduces a breaking change

- [ ] Yes
- [x] No

### CODE REVIEW

File: notebooks/data-sources/TestGrid/metrics/metric_template.ipynb [hunk 1]
```diff
@@ -315,9 +316,16 @@
...
     "                            # for all the other tests in each grid\n",
     "                            if tests[i] == \"Overall\":\n",
-    "                                test_duration = graphs[i][0][\"values\"][0][j]\n",
-    "                            else:\n",
-    "                                test_duration = None\n",
+    "                                try:\n",
+    "                                    test_duration = graphs[i][0][\"values\"][0][j]\n",
+    "                                except IndexError:\n",
+    "                                    print(\n",
+    "                                        \"Exception occurred in test_duration calculation\",\n",
```
Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/metric_template.ipynb:323]: Can you provide a more descriptive statement? like "No values present to calculate test duration" or something? The current comment sounds a little vague, and unclear if the user should be concerned  by exception.

File: notebooks/data-sources/TestGrid/metrics/metric_template.ipynb [hunk 2]
```diff
@@ -339,7 +347,15 @@
     "                            # we are only considering the 'Overall' test and fetching the time duration\n",
     "                            # values for this test in each grid\n",
     "                            if tests[i] == \"Overall\":\n",
-    "                                test_duration = graphs[i][0][\"values\"][0][j]\n",
+    "                                try:\n",
+    "                                    test_duration = graphs[i][0][\"values\"][0][j]\n",
+    "                                except IndexError:\n",
+    "                                    print(\n",
+    "                                        \"Exception occurred in test_duration calculation\",\n",
```
Comment (MichaelClifford) [inline notebooks/data-sources/TestGrid/metrics/metric_template.ipynb:354]: same comment as above

**Review by MichaelClifford — CHANGES_REQUESTED** [review]
:+1: Good job. Minor comment.

File: notebooks/data-sources/TestGrid/metrics/metric_template.ipynb (hunk 1 above)
Comment (oindrillac) [reply notebooks/data-sources/TestGrid/metrics/metric_template.ipynb:323]: Yes, will do 👍

File: notebooks/data-sources/TestGrid/metrics/metric_template.ipynb (hunk 2 above)
Comment (oindrillac) [reply notebooks/data-sources/TestGrid/metrics/metric_template.ipynb:354]: 👍

File: notebooks/data-sources/TestGrid/metrics/metric_template.ipynb (hunk 1 above)
Comment (oindrillac) [reply notebooks/data-sources/TestGrid/metrics/metric_template.ipynb]: made the change!

**Comment by oindrillac** [comment]
@MichaelClifford sorry, can you re approve. just fixed the pre-commit

**Comment by oindrillac** [comment]
@Shreyanand yes, that makes sense. lets address that in a new pr?

**Pull request merged**